compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Validate.hs

```diff
@@ -0,0 +1,75 @@
+-- Copyright (c) 2020 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+module DA.Daml.Compiler.Validate
+  ( Validation
+  , runValidation
+  , validateDar
+  ) where
+
+import Control.Lens (toListOf)
+import DA.Daml.LF.Ast.Optics (packageRefs)
+import Data.List.Extra ((\\), nubSort)
+import System.FilePath (isExtensionOf)
+import qualified "zip-archive" Codec.Archive.Zip as ZipArchive
+import qualified DA.Daml.LF.Ast as LF
+import qualified DA.Daml.LF.Proto3.Archive as Archive
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.ByteString.UTF8 as BSUTF8
+import qualified Data.Text.Extended as T
+
+-- | Monad for performing Validation
+newtype Validation a = Validation (Either ValidationError a) deriving (Functor,Applicative,Monad)
+
+data ValidationError
+  = VeArchiveError FilePath Archive.ArchiveError
+  | VeUnclosed [LF.PackageId]
+
+validationError :: ValidationError -> Validation a
+validationError = Validation . Left
+
+-- | Runs a Validation, throwing an error when the validation fails
+runValidation :: Validation a -> a
+runValidation = \case
+  Validation (Left e) -> error (show e)
+  Validation (Right a) -> a
+
+instance Show ValidationError where
+  show = \case
+    VeArchiveError fp err -> unlines
+      [ "Invalid DAR."
+      , "DALF entry cannot be decoded: " <> fp
+      , show err ]
+    VeUnclosed pids -> unlines $
+      [ "Invalid DAR."
+      , "The DAR is not closed."
+      , "The following packages are referenced but not defined:"]
+      ++ ["- " <> T.unpack (LF.unPackageId pid) | pid <- pids ]
+
+-- | Validate the byte string of a loaded DAR: that all DALFs are well-typed and the DAR is closed over packages
+validateDar :: BSUTF8.ByteString -> Validation Int
+validateDar bytes = do
+  let dar = ZipArchive.toArchive $ BSL.fromStrict bytes
+  let entries = [e | e <- ZipArchive.zEntries dar, ".dalf" `isExtensionOf` ZipArchive.eRelativePath e]
+  (definedIds,packages) <- unzip <$> mapM validateEntry entries
+  -- TODO: validate the main DALF package exists
+  validateClosedPackages definedIds packages
+  mapM_ validateWellTyped packages
+  return (length packages)
+
+validateEntry :: ZipArchive.Entry -> Validation (LF.PackageId,LF.Package)
+validateEntry entry = do
+  let bs = BSL.toStrict $ ZipArchive.fromEntry entry
+  case Archive.decodeArchive Archive.DecodeAsMain bs of
+    Left err -> validationError $ VeArchiveError (ZipArchive.eRelativePath entry) err
+    Right x -> return x
+
+validateClosedPackages :: [LF.PackageId] -> [LF.Package] -> Validation ()
```

Comment:
You don’t need this check. Just running the typechecker will enforce this.

Reply:
done